### ISSUE TITLE
match do catch-22

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#44087]
It seems that you need to add parentheses in order to put a `catch` expression in a match head, but if you do, [you get a warning](https://play.rust-lang.org/?gist=e9a2d7c01142bb162cffa2128b1dcafb&version=nightly):

```rust
#![feature(catch_expr)]

fn main() {
    match (do catch { }) { } //~WARN unnecessary parentheses around `match` head expression
    match do catch { } { } //~ERROR expected expression, found reserved keyword `do`
}
```

Why are the parens required anyway?

**Comment by kennytm** [comment]
I think the original idea is to make this unambiguous?

```rust
let catch = 5;
match catch { ...
```

**Comment by durka** [comment]
`catch` is already ambiguous with struct init syntax, I thought that's why we added `do`.

**Comment by shepmaster** [comment]
> catch is already ambiguous with struct init syntax,

But struct init syntax is disallowed in a `match` head.

**Comment by durka** [comment]
Then there's no ambiguity anyway. I don't understand the point of banning this.

**Comment by durka** [comment]
This somehow got even worse: now `match (do catch {}) {}` error saying the new syntax is `try`, but obviously that doesn't compile either without enabling the new edition.

(And in E2018, `match (try {}) {}` still has a warning telling you remove parentheses that you can't remove.)

**Comment by kennytm** [comment]
Given #54885 I'm labeling this a diagnostic problem.

**Comment by Mark-Simulacrum** [comment]
Updated code, still lints on the unused parens but errors without them.

```
#![feature(try_blocks)]

fn main() {
    match (try { Ok::<_, ()>(0) }) {
        Ok(_) => {},
        Err(()) => {}
    }
}
```

**Comment by bjorn3** [comment]
This now compiles:

```rust
#![feature(try_blocks)]

fn main() {
    match try { Ok::<_, ()>(0) } {
        Ok(_) => {},
        Err(()) => {}
    }
}
```
